File: src/precice/bindings/matlab/+precice/@SolverInterface/SolverInterface.m
classdef SolverInterface < handle
    %SOLVERINTERFACE Matlab wrapper for the C++ SolverInterface
    %   This class will serve as a gateway for the precice Solver Interface
    %   in C++. It has access to a private mex Function written in the new
    %   MATLAB C++ API. Such a function is actually an instance of a C++ 
    %   class that is stored internally by MATLAB. Whenever they are 
    %   "called", a subroutine of the object (operator ()) is invoked which
    %   then has access to the members of the class.
    %   Hence, we can store the actual solver interface in this class and
    %   access it by invoking the mex function.
    
    properties(Access=private)
        % Possible changes:
        % - Allow the creation of multiple SolverInterfaces. For this, we
        % should replace the interface pointer in the Gateway by a list of 
        % interface pointers and add an InterfaceID as property of this
        % class.
    end
    
    methods
        %% Construction and configuration
        % Constructor
        function obj = SolverInterface(SolverName)
            %SOLVERINTERFACE Construct an instance of this class
            if ischar(SolverName)
                SolverName = string(SolverName);
            end
            preciceGateway(uint8(0),SolverName);
        end
        
        % Destructor
        function delete(obj)
            % Delete the mex host
            preciceGateway(uint8(1));
        end
        
        % configure
        function configure(obj,configFileName)
            if ischar(configFileName)
                configFileName = string(configFileName);
            end
            preciceGateway(uint8(2),configFileName);
        end
        
        %% Steering methods
        % initialize
        function dt = initialize(obj)
            dt = preciceGateway(uint8(10));
        end
        
        % initialize Data
        function initializeData(obj)
            preciceGateway(uint8(11));
        end
        
        % advance
        function dt = advance(obj,dt)
            dt = preciceGateway(uint8(12),dt);
        end
        
        % finalize
        function finalize(obj)
            preciceGateway(uint8(13));
        end
        
        %% Status queries
        % getDimensions
        function dims = getDimensions(obj)
            dims = preciceGateway(uint8(20));
        end
        
        % isCouplingOngoing
        function bool = isCouplingOngoing(obj)
            bool = preciceGateway(uint8(21));
        end
        
        % isReadDataAvailable
        function bool = isReadDataAvailable(obj)
            bool = preciceGateway(uint8(22));
        end
        
        % isWriteDataRequired
        function bool = isWriteDataRequired(obj,dt)
            bool = preciceGateway(uint8(23),dt);
        end
        
        % isTimestepComplete
        function bool = isTimestepComplete(obj)
            bool = preciceGateway(uint8(24));
        end
        
        % hasToEvaluateSurrogateModel
        function bool = hasToEvaluateSurrogateModel(obj)
            bool = preciceGateway(uint8(25));
        end
        
        % hasToEvaluateFineModel
        function bool = hasToEvaluateFineModel(obj)
            bool = preciceGateway(uint8(25));
        end
        
        %% Action Methods
        % isActionRequired
        function bool = isActionRequired(obj,action)
            if ischar(action)
                action = string(action);
            end
            bool = preciceGateway(uint8(30),action);
        end
        
        % fulfilledAction
        function fulfilledAction(obj,action)
            if ischar(action)
                action = string(action);
            end
            preciceGateway(uint8(31),action);
        end
        
        %% Mesh Access
        % hasMesh
        function bool = hasMesh(obj,meshName)
            if ischar(meshName)
                meshName = string(meshName);
            end
            bool = preciceGateway(uint8(40),meshName);
        end
        
        % getMeshID
        function id = getMeshID(obj,meshName)
            if ischar(meshName)
                meshName = string(meshName);
            end
            id = preciceGateway(uint8(41),meshName);
        end
        
        % getMeshIDs
        function ids = getMeshIDs(obj)
            ids = preciceGateway(uint8(42));
        end
        
        % getMeshHandle not yet implemented
        
        % setMeshVertex
        function vertexId = setMeshVertex(obj,meshID,position)
            vertexId = preciceGateway(uint8(44),int32(meshID),position);
        end
        
        % getMeshVertexSize
        function vertexId = getMeshVertexSize(obj,meshID)
            vertexId = preciceGateway(uint8(45),int32(meshID));
        end
        
        % setMeshVertices
        function vertexIds = setMeshVertices(obj,meshID,inSize,positions)
            if size(positions,2) ~= inSize
                error('Number of columns in position vector must match size!');
            end
            vertexIds = preciceGateway(uint8(46),int32(meshID),int32(inSize),positions);
        end
        
        % getMeshVertices
        function positions = getMeshVertices(obj,meshID,inSize,vertexIds)
            if size(ids,2) ~= inSize
                error('Number of columns in position vector must match size!');
            end
            positions = preciceGateway(uint8(47),int32(meshID),int32(inSize),vertexIds);
        end
        
        % getMeshVertexIDsFromPositions
        function vertexIds = getMeshVertexIDsFromPositions(obj,meshID,inSize,positions)
            if size(positions,2) ~= inSize
                error('Number of columns in position vector must match size!');
            end
            vertexIds = preciceGateway(uint8(48),int32(meshID),int32(inSize),positions);
        end
        
        % setMeshEdge
        function edgeID = setMeshEdge(obj, meshID, firstVertexID, secondVertexID)
            edgeID = preciceGateway(uint8(49),int32(meshID),int32(firstVertexID),int32(secondVertexID));
        end
        
        % setMeshTriangle
        function setMeshTriangle(obj, meshID, firstEdgeID, secondEdgeID, thirdEdgeID)
            preciceGateway(uint8(50),int32(meshID),int32(firstEdgeID),int32(secondEdgeID),int32(thirdEdgeID));
        end
        
        % setMeshTriangleWithEdges
        function setMeshTriangleWithEdges(obj, meshID, firstVertexID, secondVertexID, thirdVertexID)
            preciceGateway(uint8(51),int32(meshID),int32(firstVertexID),int32(secondVertexID),int32(thirdVertexID));
        end
        
        % setMeshQuad
        function setMeshQuad(obj, meshID, firstEdgeID, secondEdgeID, thirdEdgeID, fourthEdgeID)
            preciceGateway(uint8(52),int32(meshID),int32(firstEdgeID),int32(secondEdgeID),int32(thirdEdgeID),int32(fourthEdgeID));
        end
        
        % setMeshQuadWithEdges
        function setMeshQuadWithEdges(obj, meshID, firstVertexID, secondVertexID, thirdVertexID, fourthVertexID)
            preciceGateway(uint8(53),int32(meshID),int32(firstVertexID),int32(secondVertexID),int32(thirdVertexID),int32(fourthVertexID));
        end
        
        %% Data Access
        % hasDataID
        function bool = hasData(obj,dataName,meshID)
            if ischar(dataName)
                dataName = string(dataName);
            end
            bool = preciceGateway(uint8(60),dataName,int32(meshID));
        end
        
        % getDataID
        function id = getDataID(obj,dataName,meshID)
            if ischar(dataName)
                dataName = string(dataName);
            end
            id = preciceGateway(uint8(61),dataName,int32(meshID));
        end
        
        % mapReadDataTo
        function mapReadDataTo(obj,meshID)
            preciceGateway(uint8(62),int32(meshID));
        end
        
        % mapWriteDataFrom
        function mapWriteDataFrom(obj,meshID)
            preciceGateway(uint8(63),int32(meshID));
        end
        
        % writeBlockVectorData
        function writeBlockVectorData(obj,dataID,inSize,valueIndices,values)
            if ~isa(valueIndices,'int32')
                warning('valueIndices should be allocated as int32 to prevent copying.');
                valueIndices = int32(valueIndices);
            end
            preciceGateway(uint8(64),int32(dataID),int32(inSize),valueIndices,values);
        end
        
        % writeVectorData
        function writeVectorData(obj,dataID,valueIndex,value)
            preciceGateway(uint8(65),int32(dataID),int32(valueIndex),value);
        end
        
        % writeBlockScalarData
        function writeBlockScalarData(obj,dataID,inSize,valueIndices,values)
            if ~isa(valueIndices,'int32')
                warning('valueIndices should be allocated as int32 to prevent copying.');
                valueIndices = int32(valueIndices);
            end
            preciceGateway(uint8(66),int32(dataID),int32(inSize),valueIndices,values);
        end
        
        % writeScalarData
        function writeScalarData(obj,dataID,valueIndex,value)
            preciceGateway(uint8(67),int32(dataID),int32(valueIndex),value);
        end
        
        % readBlockVectorData
        function values = readBlockVectorData(obj,dataID,inSize,valueIndices)
            if ~isa(valueIndices,'int32')
                warning('valueIndices should be allocated as int32 to prevent copying.');
                valueIndices = int32(valueIndices);
            end
            values = preciceGateway(uint8(68),int32(dataID),int32(inSize),valueIndices);
        end
        
        % readVectorData
        function value = readVectorData(obj,dataID,valueIndex)
            value = preciceGateway(uint8(69),int32(dataID),int32(valueIndex));
        end
        
        % readBlockScalarData
        function values = readBlockScalarData(obj,dataID,inSize,valueIndices,transpose)
            if nargin<5
                transpose=false;
            end
            if ~isa(valueIndices,'int32')
                warning('valueIndices should be allocated as int32 to prevent copying.');
                valueIndices = int32(valueIndices);
            end
            values = preciceGateway(uint8(70),int32(dataID),int32(inSize),valueIndices,transpose);
        end
        
        % readScalarData
        function value = readScalarData(obj,dataID,valueIndex)
            value = preciceGateway(uint8(71),int32(dataID),int32(valueIndex));
        end
    end
end

File: src/precice/bindings/python_future/test/SolverInterface.cpp
#include "precice/SolverInterface.hpp"
#include <iostream>
#include <numeric>
#include <cassert>

std::vector<double> fake_read_write_buffer;
int fake_dimensions;
int fake_mesh_id;
std::vector<int> fake_ids;
int n_fake_vertices;
std::string fake_data_name;
int fake_data_id;

namespace precice {

namespace impl{
class SolverInterfaceImpl{};
}

SolverInterface:: SolverInterface
(
  const std::string& participantName,
  int                solverProcessIndex,
  int                solverProcessSize )
{
  fake_read_write_buffer = std::vector<double>();
  fake_dimensions = 3;
  fake_mesh_id = 0;
  fake_data_id = 15;
  fake_data_name = "FakeData";
  n_fake_vertices = 3;
  fake_ids.resize(n_fake_vertices);
  std::iota(fake_ids.begin(), fake_ids.end(), 0);
}

SolverInterface::~SolverInterface() = default;

void SolverInterface:: configure
(
  const std::string& configurationFileName )
{}

double SolverInterface:: initialize(){return -1;}

void SolverInterface:: initializeData()
{}

double SolverInterface:: advance
(
  double computedTimestepLength )
{return -1;}

void SolverInterface:: finalize()
{}

int SolverInterface:: getDimensions() const
{return fake_dimensions;}

bool SolverInterface:: isCouplingOngoing() const
{
  return 0;
}

bool SolverInterface:: isReadDataAvailable() const
{
  return 0;
}

bool SolverInterface:: isWriteDataRequired
(
  double computedTimestepLength ) const
{
  return 0;
}

bool SolverInterface:: isTimestepComplete() const
{
  return 0;
}

bool SolverInterface:: isActionRequired
(
  const std::string& action ) const
{
  return 0;
}

void SolverInterface:: fulfilledAction
(
  const std::string& action )
{}

bool SolverInterface:: hasMesh
(
  const std::string& meshName ) const
{
  return 0;
}

int SolverInterface:: getMeshID
(
  const std::string& meshName ) const
{
  return fake_mesh_id;
}

std::set<int> SolverInterface:: getMeshIDs() const
{
  return std::set<int>();
}

bool SolverInterface:: hasData
(
  const std::string& dataName, int meshID ) const
{
  return 0;
}

int SolverInterface:: getDataID
(
  const std::string& dataName, int meshID ) const
{
  if(meshID == fake_mesh_id && dataName == fake_data_name)
  {
    return fake_data_id;
  }
  else
  {
    return -1;
  }
}

bool SolverInterface::hasToEvaluateSurrogateModel() const
{
  return 0;
}

bool SolverInterface::hasToEvaluateFineModel() const
{
  return 0;
}

int SolverInterface:: setMeshVertex
(
  int           meshID,
  const double* position )
{
  return 0;
}

int SolverInterface:: getMeshVertexSize
(
  int meshID) const
{
  return n_fake_vertices;
}

void SolverInterface:: setMeshVertices
(
  int           meshID,
  int           size,
  const double* positions,
  int*          ids )
{
  assert (size == fake_ids.size());
  std::copy(fake_ids.begin(), fake_ids.end(), ids);
}

void SolverInterface:: getMeshVertices
(
  int        meshID,
  int        size,
  const int* ids,
  double*    positions ) const
{
  for(int i = 0; i < size; i++){
      positions[fake_dimensions * i] = i;
      positions[fake_dimensions * i + 1] = i + n_fake_vertices;
      positions[fake_dimensions * i + 2] = i + 2 * n_fake_vertices;
  }
}

void SolverInterface:: getMeshVertexIDsFromPositions
(
  int           meshID,
  int           size,
  const double* positions,
  int*          ids ) const
{
  assert (size == fake_ids.size());
  std::copy(fake_ids.begin(), fake_ids.end(), ids);
}

int SolverInterface:: setMeshEdge
(
  int meshID,
  int firstVertexID,
  int secondVertexID )
{
  return -1;
}

void SolverInterface:: setMeshTriangle
(
  int meshID,
  int firstEdgeID,
  int secondEdgeID,
  int thirdEdgeID )
{}

void SolverInterface:: setMeshTriangleWithEdges
(
  int meshID,
  int firstVertexID,
  int secondVertexID,
  int thirdVertexID )
{}

void SolverInterface:: setMeshQuad
(
  int meshID,
  int firstEdgeID,
  int secondEdgeID,
  int thirdEdgeID,
  int fourthEdgeID )
{}

void SolverInterface:: setMeshQuadWithEdges
(
  int meshID,
  int firstVertexID,
  int secondVertexID,
  int thirdVertexID,
  int fourthVertexID )
{}

void SolverInterface:: mapReadDataTo
(
  int toMeshID )
{}

void SolverInterface:: mapWriteDataFrom
(
  int fromMeshID )
{}


void SolverInterface:: writeBlockVectorData
(
  int     dataID,
  int     size,
  const int*    valueIndices,
  const double* values )
{
  fake_read_write_buffer.clear();
  for(int i = 0; i < size * this->getDimensions(); i++){
      fake_read_write_buffer.push_back(values[i]); 
    }
}

void SolverInterface:: writeVectorData
(
  int           dataID,
  int           valueIndex,
  const double* value )
{
  fake_read_write_buffer.clear();
  for(int i = 0; i < this->getDimensions(); i++){
      fake_read_write_buffer.push_back(value[i]); 
    }
}

void SolverInterface:: writeBlockScalarData
(
  int           dataID,
  int           size,
  const int*    valueIndices,
  const double* values )
{
  fake_read_write_buffer.clear();
  for(int i = 0; i < size; i++){
      fake_read_write_buffer.push_back(values[i]); 
    }
}

void SolverInterface:: writeScalarData
(
  int    dataID,
  int    valueIndex,
  double value )
{
    fake_read_write_buffer.clear();
    fake_read_write_buffer.push_back(value); 
}

void SolverInterface:: readBlockVectorData
(
  int        dataID,
  int        size,
  const int* valueIndices,
  double*    values ) const
{
  for(int i = 0; i < size * this->getDimensions(); i++){
      values[i] = fake_read_write_buffer[i];
    }
}

void SolverInterface:: readVectorData
(
  int     dataID,
  int     valueIndex,
  double* value ) const
{
  for(int i = 0; i < this->getDimensions(); i++){
      value[i] = fake_read_write_buffer[i];
    }
}

void SolverInterface:: readBlockScalarData
(
  int        dataID,
  int        size,
  const int* valueIndices,
  double*    values ) const
{
  for(int i = 0; i<size; i++){
      values[i] = fake_read_write_buffer[i];
    }
}

void SolverInterface:: readScalarData
(
  int     dataID,
  int     valueIndex,
  double& value ) const
{
    value = fake_read_write_buffer[0]; 
}

namespace constants {

const std::string& actionWriteInitialData()
{
    static std::string dummy ("dummy");
    return dummy;
}

const std::string& actionWriteIterationCheckpoint()
{
    static std::string dummy ("dummy");
    return dummy;
}

const std::string& actionReadIterationCheckpoint()
{
    static std::string dummy ("dummy");
    return dummy;
}

} // namespace precice, constants

} // namespace precice
